File: src/main/java/com/vc/medium/L22.scala
package com.vc.medium

object L22 {
  def generateParenthesis(n: Int): List[String] = {
    import scala.collection.mutable
    val list = new mutable.ListBuffer[String]()
    def solve(s: String, open: Int, close: Int): Unit = {
      if(s.length == n * 2) list += s
      else {
        if(open < n)
          solve(s + "(", open + 1, close)
        if(close < open)
          solve(s + ")", open, close + 1)
      }
    }
    solve("", 0, 0)
    list.toList
  }
}

File: src/main/java/com/vc/medium/L22.java
package com.vc.medium;

import java.util.*;

class L22 {
    private List<String> list;
    private int n = 0;

    public List<String> generateParenthesis(int n) {
        this.list = new ArrayList<>();
        this.n = n;
        helper(0, 0, "");
        return list;
    }

    private void helper(int open, int close, String str) {
        if(open == close && close == n) list.add(str);

        if(open > close)
            helper(open, close + 1, str + ")");
        if(open < n)
            helper(open + 1, close, str + "(");
    }
}